sweepbatcher/sweep_batch.go

```diff
@@ -0,0 +1,1369 @@
+	)
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	b.wg.Add(1)
+	go func() {
+		defer cancel()
+		defer b.wg.Done()
+
+		for {
+			select {
+			case conf := <-confChan:
+				select {
+				case b.confChan <- conf:
+
+				case <-ctx.Done():
+				}
+				return
+
+			case err := <-errChan:
+				b.writeToErrChan(err)
+				return
+
+			case <-reorgChan:
+				// A re-org has been detected. We set the batch
+				// state back to open since our batch
+				// transaction is no longer present in any
+				// block. We can accept more sweeps and try to
+				// publish new transactions, at this point we
+				// need to monitor again for a new spend.
+				select {
+				case b.reorgChan <- struct{}{}:
+				case <-ctx.Done():
+				}
+				return
+
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// handleSpend handles a spend notification.
+func (b *batch) handleSpend(ctx context.Context, spendTx *wire.MsgTx) error {
+	var (
+		txHash     = spendTx.TxHash()
+		purgeList  = make([]SweepRequest, 0, len(b.sweeps))
+		notifyList = make([]sweep, 0, len(b.sweeps))
+	)
+	b.batchTxid = &txHash
+	b.batchPkScript = spendTx.TxOut[0].PkScript
+
+	// As a previous version of the batch transaction may get confirmed,
+	// which does not contain the latest sweeps, we need to detect the
+	// sweeps that did not make it to the confirmed transaction and feed
+	// them back to the batcher. This will ensure that the sweeps will enter
+	// a new batch instead of remaining dangling.
+	for _, sweep := range b.sweeps {
+		found := false
+
+		for _, txIn := range spendTx.TxIn {
+			if txIn.PreviousOutPoint == sweep.outpoint {
+				found = true
+				notifyList = append(notifyList, sweep)
+			}
+		}
+
+		// If the sweep's outpoint was not found in the transaction's
+		// inputs this means it was left out. So we delete it from this
+		// batch and feed it back to the batcher.
+		if !found {
+			newSweep := sweep
+			delete(b.sweeps, sweep.swapHash)
+			purgeList = append(purgeList, SweepRequest{
+				SwapHash: newSweep.swapHash,
+				Outpoint: newSweep.outpoint,
+				Value:    newSweep.value,
+				Notifier: newSweep.notifier,
+			})
+		}
+	}
+
+	for _, sweep := range notifyList {
+		// Save the sweep as completed.
+		err := b.persistSweep(ctx, sweep, true)
+		if err != nil {
+			return err
+		}
+
+		// If the sweep's notifier is empty then this means that a swap
+		// is not waiting to read an update from it, so we can skip
+		// the notification part.
+		if sweep.notifier == nil ||
+			*sweep.notifier == (SpendNotifier{}) {
+
+			continue
+		}
+
+		// Dispatch the sweep notifier, we don't care about the outcome
+		// of this action so we don't wait for it.
+		go notifySweepSpend(ctx, sweep, spendTx)
+	}
+
+	// If we have sweeps to purge, we send them back to the batcher.
+	for _, sweep := range purgeList {
+		sweep := sweep
+		// We don't care for the outcome of the purge, so we don't wait
+		// for it.
+		go func() {
```

Comment:
nit: could do the whole purge list in a single goroutine.

Reply:
oh! nice one, ty